monai/inferers/splitter.py

```diff
@@ -58,142 +91,372 @@ def __call__(self, inputs: Any) -> Iterable[tuple[torch.Tensor, Sequence[int]]]:
+        )
+        return inputs[slices]
+
+    def get_input_shape(self, inputs: Any) -> tuple:
+        """
+        Return the input spatial shape.
+
+        Args:
+            inputs: either a tensor of shape BCHW[D], representing a batch of images,
+                or a filename (str) or list of filenames to the image(s).
+
+        Returns:
+            spatial_shape
+        """
+        return tuple(inputs.shape[self.non_spatial_ndim :])
+
+    def get_output_shape(self, inputs: Any) -> tuple:
+        """
+        Return the actual spatial shape covered by the output split patches.
+        For instance, if the input image is padded, the actual spatial shape will be enlarged
+        and not the same as input spatial shape.
+
+        Args:
+            inputs: either a tensor of shape BCHW[D], representing a batch of images,
+                or a filename (str) or list of filenames to the image(s).
+
+        Returns:
+            padded_spatial_shape
+
+        """
+        spatial_shape = self.get_input_shape(inputs)
+        if not self.pad_mode:
+            return spatial_shape
+        spatial_ndim = len(spatial_shape)
+        patch_size, overlap, offset = self._get_valid_shape_parameters(spatial_shape)
+        pad_size, _ = self._calculate_pad_size(spatial_shape, spatial_ndim, patch_size, offset, overlap)
+        padded_spatial_shape = tuple(ss + ps + pe for ss, ps, pe in zip(spatial_shape, pad_size[1::2], pad_size[::2]))
+
+        return padded_spatial_shape
+
+    def __call__(self, inputs: Any) -> Iterable[tuple[torch.Tensor, Sequence[int]]]:
         """Split the input tensor into patches and return patches and locations.
 
         Args:
-            inputs: a torch.Tensor with BCHW[D] dimensions, representing an image or a batch of images.
+            inputs: either a torch.Tensor with BCHW[D] dimensions, representing an image or a batch of images
 
         Yields:
             tuple[torch.Tensor, Sequence[int]]: yields tuple of patch and location
         """
-        n_non_spatial_dims = 2
-        spatial_ndim = inputs.ndim - n_non_spatial_dims
-        spatial_shape = inputs.shape[n_non_spatial_dims:]
-        patch_size = self._get_valid_patch_size(spatial_ndim)
-        overlap = self._get_valid_overlap(spatial_ndim, patch_size)
-        offset = self._get_valid_offset(spatial_shape, spatial_ndim, patch_size)
+
+        if not isinstance(inputs, torch.Tensor):
+            raise ValueError(f"The input should be a tensor. {type(inputs)} is given.")
+
+        spatial_shape = inputs.shape[self.non_spatial_ndim :]
+        spatial_ndim = len(spatial_shape)
+        patch_size, overlap, offset = self._get_valid_shape_parameters(spatial_shape)
         pad_size, is_start_padded = self._calculate_pad_size(spatial_shape, spatial_ndim, patch_size, offset, overlap)
 
-        if any(pad_size):
+        # Padding
+        if self.pad_mode and any(pad_size):
```

Comment:
Maybe we could move these padding part in to `get_output_shape`, seems there are some duplicated code. What do you think?

Reply:
I know but they were required to satisfy the logic for cropping. I will check if I can change it considering the revision in the cropping. thanks

Reply:
Will this part be updated or not?

Reply:
I checked that and I think we should keep this duplicity as is. The reason for that is `get_output_shape` (now called `get_padded_shape`) is being called in a different place than `__call__`. In other words, `__call__` is being called by any user/code that uses this Splitter but `get_padded_shape` only needed for specific situations where a user/code wants to know the padded shape of the input, for instance in `PatchInferer` where we are interested to match the output merged shape with the original input shape.

Reply:
Thanks, mark it as resolved now.